src/OpenLoco/Objects/ObjectManager.cpp

```diff
@@ -77,10 +79,371 @@ namespace OpenLoco::ObjectManager
+        std::memcpy(&entryBuffer[newEntrySize], &objHeader3, sizeof(objHeader3));
+        newEntrySize += sizeof(objHeader3);
+
+        auto* ptr = &_112A17F[0];
+
+        // ObjectList1 (required objects)
+        const uint8_t size1 = uint8_t(*ptr++);
+        entryBuffer[newEntrySize++] = std::byte(size1);
+        std::memcpy(&entryBuffer[newEntrySize], ptr, sizeof(ObjectHeader) * size1);
+        newEntrySize += sizeof(ObjectHeader) * size1;
+        ptr += sizeof(ObjectHeader) * size1;
+
+        // ObjectList2 (also loads objects {used for category selection})
+        const uint8_t size2 = uint8_t(*ptr++);
+        entryBuffer[newEntrySize++] = std::byte(size2);
+        std::memcpy(&entryBuffer[newEntrySize], ptr, sizeof(ObjectHeader) * size2);
+        newEntrySize += sizeof(ObjectHeader) * size2;
+        ptr += sizeof(ObjectHeader) * size2;
+
+        return std::make_pair(entry, newEntrySize);
+    }
+
+    static void addObjectToIndex(const fs::path filepath, size_t& usedBufferSize)
+    {
+        ObjectHeader objHeader{};
+        {
+            std::ifstream stream;
+            stream.open(filepath, std::ios::in | std::ios::binary);
+            Utility::readData(stream, objHeader);
+            if (stream.gcount() != sizeof(objHeader))
+            {
+                return;
+            }
+        }
+
+        const auto curObjPos = usedBufferSize;
+        const auto partialNewEntry = createPartialNewEntry(&_installedObjectList[usedBufferSize], objHeader, filepath.filename());
+        usedBufferSize += partialNewEntry.second;
+        _installedObjectCount++;
+
+        _isPartialLoaded = true;
+        _50D158 = _112A17F;
+        getScenarioText(objHeader);
+        _50D158 = reinterpret_cast<std::byte*>(-1);
+        _isPartialLoaded = false;
+        if (_installedObjectCount == 0)
+        {
+            return;
+        }
+        _installedObjectCount--;
+
+        // Rewind as it is only a partial object loaded
+        usedBufferSize = curObjPos;
+
+        // Load full entry into temp buffer.
+        // 0x009D1CC8
+        std::byte newEntryBuffer[0x2000];
+        const auto [newEntry, newEntrySize] = createNewEntry(newEntryBuffer, objHeader, filepath.filename());
+
+        freeScenarioText();
+
+        auto* indexPtr = *_installedObjectList;
+        for (uint32_t i = 0; i < _installedObjectCount; i++)
+        {
+            auto entry = ObjectIndexEntry::read(&indexPtr);
+            if (strcmp(entry._name, newEntry._name) < 0)
+            {
+                break;
+            }
+        }
+
+        auto moveSize = usedBufferSize - (indexPtr - *_installedObjectList);
+        std::memmove(indexPtr + newEntrySize, indexPtr, moveSize);
+        std::memcpy(indexPtr, newEntryBuffer, newEntrySize);
+        usedBufferSize += newEntrySize;
+
+        _installedObjectCount++;
+    }
+
+    // 0x0047118B
+    static void createIndex(const ObjectFolderState& currentState)
```

Comment:
This function is very dense right now. Could you add a bit more whitespace and/or comments for readability?